pyi.py

```diff
@@ -442,6 +515,52 @@ def _visit_function(self, node: ast.FunctionDef | ast.AsyncFunctionDef) -> None:
                     continue
             self.error(statement, Y010)
 
+        if self.in_class:
+            self.check_self_typevars(node)
+
+    def check_self_typevars(self, node: ast.FunctionDef | ast.AsyncFunctionDef) -> None:
+        pos_or_keyword_args = node.args.args
+
+        if not pos_or_keyword_args:
+            return
+        return_annotation = node.returns
+
+        if not isinstance(return_annotation, ast.Name):
+            return
+        first_arg_annotation = pos_or_keyword_args[0].annotation
+
+        if not isinstance(first_arg_annotation, (ast.Name, ast.Subscript)):
+            return
+
+        if node.name == "__new__":
+            self._check_class_method_for_bad_typevars(
+                method=node,
+                first_arg_annotation=first_arg_annotation,
+                return_annotation=return_annotation,
+            )
+            return
+
+        decorator_names = {
+            decorator.id
+            for decorator in node.decorator_list
+            if isinstance(decorator, ast.Name)
+        }
+
+        if "classmethod" in decorator_names:
```

Comment:
Maybe this should be: `if "classmethod" in decorator_names or node.name == "__new__":`

The intent is to treat `__new__` as a classmethod, and it would be much more clear if the two checks were combined.